rook: Handling gosec validation error

nizamial09 opened the pull request:
Fixed G101: Potential hardcoded credential variable by renaming certain variables
Fixed G104: Unhandled Errors are handled now
Fixed G303: Poor file permissions used for directory by using the most narrow permission
Fixed G304: Potential File inclusion by cleaning up the file path


Signed-off-by: Nizamudeen <nia@redhat.com>

<!-- Please take a look at our [Contributing](https://rook.io/docs/rook/master/development-flow.html)
documentation before submitting a Pull Request!
Thank you for contributing to Rook! -->

**Description of your changes:**

Summary of gosec after making the changes. 
```Summary:
   Files: 380
   Lines: 50900
   Nosec: 7
  Issues: 9
```
**Which issue is resolved by this Pull Request:**
Resolves #3765, #4579,#4580

**Checklist:**

- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [x] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [ ] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

[test full]

nizamial09 commented:
@travisn pls review this one.

leseb commented on pkg/daemon/ceph/agent/agent.go:
```diff
@@ -113,7 +113,9 @@ func (a *Agent) Run() error {
 	stopChan := make(chan struct{})
-	clusterController.StartWatch(v1.NamespaceAll, stopChan)
+	if err := clusterController.StartWatch(v1.NamespaceAll, stopChan); err != nil {
+		logger.Errorf("Cluster Controller failed to start watch")
```
Share the `err`

leseb commented on pkg/daemon/ceph/agent/flexvolume/server.go:
```diff
@@ -89,7 +90,9 @@ func (s *FlexvolumeServer) Start(driverVendor, driverName string) error {
 		logger.Info("Deleting unix domain socket file.")
-		os.Remove(unixSocketFile)
+		if err := os.Remove(unixSocketFile); err != nil {
+			logger.Errorf("%+v", err)
```
Name the error

leseb commented on pkg/daemon/ceph/agent/flexvolume/server.go:
```diff
@@ -119,7 +123,9 @@ func (s *FlexvolumeServer) StopAll() {
 			logger.Infof("Deleting unix domain socket file %s.", unixSocketFile)
-			os.Remove(unixSocketFile)
+			if err := os.Remove(unixSocketFile); err != nil {
+				logger.Errorf("%+v", err)
```
ditto

leseb commented on pkg/operator/cassandra/controller/controller.go:
```diff
@@ -91,7 +94,9 @@ func New(
 	// logged for sample-controller types.
-	rookScheme.AddToScheme(scheme.Scheme)
+	if err := rookScheme.AddToScheme(scheme.Scheme); err != nil {
+		logger.Errorf("%+v", err)
```
ditto

leseb commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -457,7 +457,9 @@ func (c *cluster) postMonStartupActions() error {
 				// so instead of trying to active msgr2 when mons are not ready, we activate it when we believe that's the right time
-				client.EnableMessenger2(c.context, c.Namespace)
+				if err := client.EnableMessenger2(c.context, c.Namespace); err != nil {
+					logger.Errorf("%+v", err)
```
ditto

nizamial09 replied:
> Overall please name the errors instead of just print it.

For some errors I don't know the name of errors. In that cases what should I do?

leseb replied:
Well, each error is based on a task so just name that task.
Here, for example, you can simply say: `failed to enable messenger 2`.

nizamial09 replied:
Okay

BlaineEXE commented on pkg/daemon/ceph/agent/flexvolume/server.go:
```diff
@@ -193,7 +199,7 @@ func LoadFlexSettings(directory string) []byte {
 	// copying flex volume
 	if _, err := os.Stat(driverDir); os.IsNotExist(err) {
-		err := os.Mkdir(driverDir, 0755)
+		err := os.Mkdir(driverDir, 0755) // #nosec
```
Please include information in the comment about why `#nosec` is used here. It's important to note things like this in code.

BlaineEXE commented on pkg/daemon/ceph/agent/flexvolume/server.go:
```diff
@@ -246,13 +252,13 @@ func configureFlexVolume(driverFile, driverDir, driverName string) error {
 	defer srcFile.Close()
 
-	destFile, err := os.OpenFile(dest, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0755) // creates if file doesn't exist
+	destFile, err := os.OpenFile(dest, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0755) // creates if file doesn't exist #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/daemon/ceph/osd/agent.go:
```diff
@@ -519,7 +519,7 @@ func (a *OsdAgent) prepareOSD(context *clusterd.Context, cfg *osdConfig) (*oposd
 
 	if devPartInfo != nil {
-		sys.UnmountDevice(devPartInfo.pathToUnmount, context.Executor)
+		sys.UnmountDevice(devPartInfo.pathToUnmount, context.Executor) // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/daemon/ceph/osd/daemon.go:
```diff
@@ -46,7 +46,7 @@ func StartOSD(context *clusterd.Context, osdType, osdID, osdUUID, lvPath string,
 	// ensure the config mount point exists
 	configDir := fmt.Sprintf("/var/lib/ceph/osd/ceph-%s", osdID)
-	err := os.Mkdir(configDir, 0755)
+	err := os.Mkdir(configDir, 0755) // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/daemon/ceph/test/info.go:
```diff
@@ -6,6 +6,7 @@ import (
 	"path"
 )
 
+// #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/daemon/discover/discover.go:
```diff
@@ -90,7 +90,7 @@ func Run(context *clusterd.Context, probeInterval time.Duration, useCV bool) err
 	udevEvents := make(chan string)
 	go udevBlockMonitor(udevEvents, udevEventPeriod)
-
+	// #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/daemon/util/copybins.go:
```diff
@@ -67,5 +68,5 @@ func copyBinary(sourceDir, targetDir, filename string) error {
 	}
 
-	return os.Chmod(targetPath, 0755)
+	return os.Chmod(targetPath, 0755) // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/cassandra/sidecar/config.go:
```diff
@@ -417,8 +417,8 @@ func getJolokiaConfig() string {
 	var initial, override map[string]interface{}
-	yaml.Unmarshal(initialYAML, &initial)
-	yaml.Unmarshal(overrideYAML, &override)
+	yaml.Unmarshal(initialYAML, &initial)   // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/ceph/cluster/crash/crash.go:
```diff
@@ -46,7 +46,7 @@ import (
 const (
 	crashCollectorKeyringUsername = "client.crash"
-	crashCollectorSecretName      = "rook-ceph-crash-collector-keyring"
+	crashCollectorSecretName      = "rook-ceph-crash-collector-keyring" // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/ceph/cluster/mgr/dashboard.go:
```diff
@@ -39,7 +39,7 @@ const (
 	dashboardPortHTTP              = 7000
 	dashboardUsername              = "admin"
-	dashboardPasswordName          = "rook-ceph-dashboard-password"
+	dashboardPasswordName          = "rook-ceph-dashboard-password" // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/ceph/cluster/mgr/dashboard.go:
```diff
@@ -295,7 +295,7 @@ func decodeSecret(secret *v1.Secret) (string, error) {
 	logger.Infof("restarting the mgr module")
-	client.MgrDisableModule(c.context, c.Namespace, dashboardModuleName)
-	client.MgrEnableModule(c.context, c.Namespace, dashboardModuleName, true)
+	client.MgrDisableModule(c.context, c.Namespace, dashboardModuleName)      // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/ceph/cluster/mgr/spec.go:
```diff
@@ -80,8 +80,8 @@ func (c *Cluster) makeDeployment(mgrConfig *mgrConfig) *apps.Deployment {
 	}
-	c.annotations.ApplyToObjectMeta(&podSpec.ObjectMeta)
-	c.applyPrometheusAnnotations(&podSpec.ObjectMeta)
+	c.annotations.ApplyToObjectMeta(&podSpec.ObjectMeta) // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/ceph/cluster/mgr/spec.go:
```diff
@@ -135,6 +135,7 @@ func (c *Cluster) needHTTPBindFix() bool {
 // always do this and (b) make sure that all forms of the configuration option
 // are removed (see the init container factory method). Once the minimum
 // supported version of Rook contains the fix, all of this can be removed.
+// #nosec
```
Ditto on document `#nosec`

BlaineEXE replied:
It seems dangerous to disable security checking for an entire method that does as much as this one does.

BlaineEXE commented on pkg/operator/ceph/cluster/osd/status.go:
```diff
@@ -270,7 +270,7 @@ func (c *Cluster) handleStatusConfigMapStatus(nodeName string, config *provision
 				c.startOSDDaemonsOnNode(nodeName, config, configMap, status)
 			}
-			// remove the status configmap that indicated the progress
+			// remove the status configmap that indicated the progress #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/ceph/csi/secrets.go:
```diff
@@ -28,15 +28,15 @@ import (
 	csiKeyringRBDNodeUsername        = "client.csi-rbd-node"
-	csiRBDNodeSecret                 = "rook-csi-rbd-node"
-	csiRBDProvisionerSecret          = "rook-csi-rbd-provisioner"
+	csiRBDNodeSecret                 = "rook-csi-rbd-node"        // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/ceph/csi/secrets.go:
```diff
@@ -28,15 +28,15 @@ import (
 	csiKeyringCephFSNodeUsername        = "client.csi-cephfs-node"
-	csiCephFSNodeSecret                 = "rook-csi-cephfs-node"
-	csiCephFSProvisionerSecret          = "rook-csi-cephfs-provisioner"
+	csiCephFSNodeSecret                 = "rook-csi-cephfs-node"        // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/ceph/disruption/clusterdisruption/add.go:
```diff
@@ -41,7 +41,7 @@ func Add(mgr manager.Manager, context *controllerconfig.Context) error {
 	// Add the cephv1 scheme to the manager scheme
 	mgrScheme := mgr.GetScheme()
-	cephv1.AddToScheme(mgr.GetScheme())
+	cephv1.AddToScheme(mgr.GetScheme()) // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/ceph/disruption/clusterdisruption/osd.go:
```diff
@@ -273,15 +273,15 @@ func (r *ReconcileClusterDisruption) updateNoout(pdbStateMap *corev1.ConfigMap,
 			if time.Since(nooutSetTime) >= r.maintenanceTimeout {
 				// noout expired
-				osdDump.UpdateFlagOnCrushUnit(r.context.ClusterdContext, false, namespace, failureDomain, nooutFlag)
+				osdDump.UpdateFlagOnCrushUnit(r.context.ClusterdContext, false, namespace, failureDomain, nooutFlag) // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/ceph/disruption/clusterdisruption/osd.go:
```diff
@@ -273,15 +273,15 @@ func (r *ReconcileClusterDisruption) updateNoout(pdbStateMap *corev1.ConfigMap,
 			} else {
 				// set noout
-				osdDump.UpdateFlagOnCrushUnit(r.context.ClusterdContext, true, namespace, failureDomain, nooutFlag)
+				osdDump.UpdateFlagOnCrushUnit(r.context.ClusterdContext, true, namespace, failureDomain, nooutFlag) // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/ceph/disruption/clusterdisruption/osd.go:
```diff
@@ -273,15 +273,15 @@ func (r *ReconcileClusterDisruption) updateNoout(pdbStateMap *corev1.ConfigMap,
 		} else {
 			// ensure noout unset
-			osdDump.UpdateFlagOnCrushUnit(r.context.ClusterdContext, false, namespace, failureDomain, nooutFlag)
+			osdDump.UpdateFlagOnCrushUnit(r.context.ClusterdContext, false, namespace, failureDomain, nooutFlag) // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/ceph/disruption/machinedisruption/add.go:
```diff
@@ -29,7 +29,7 @@ import (
 // Add adds a new Controller to the Manager based on machinedisruption.ReconcileMachineDisruption and registers the relevant watches and handlers.
 // Read more about how Managers, Controllers, and their Watches, Handlers, Predicates, etc work here:
-// https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg
+// https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/ceph/disruption/machinelabel/add.go:
```diff
@@ -14,6 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// #nosec
```
`#nosec` on an entire package?!

BlaineEXE commented on pkg/operator/ceph/nfs/nfs.go:
```diff
@@ -152,7 +152,7 @@ func (c *CephNFSController) runGaneshaRadosGrace(nfs cephv1.CephNFS, name, actio
 	moniker := ganeshaRadosGraceCmd + " " + action + " " + nodeID
 	// Need to run a command with CEPH_CONF env var set, so don't use c.context.Executor
-	x := exec.Command(cmd, args...)
+	x := exec.Command(cmd, args...) // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/ceph/test/podspec.go:
```diff
@@ -69,7 +69,7 @@ func (ps *PodSpecTester) AssertVolumesMeetCephRequirements(
 	keyringSecretName := fmt.Sprintf("rook-ceph-%s-%s-keyring", daemonType, daemonID)
 	if daemonType == config.MonType {
-		keyringSecretName = "rook-ceph-mons-keyring" // mons share a keyring
+		keyringSecretName = "rook-ceph-mons-keyring" // mons share a keyring #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/edgefs/cluster/controller.go:
```diff
@@ -16,7 +16,7 @@ limitations under the License.
 */
 
-// Package cluster to manage a edgefs cluster.
+// Package cluster to manage a edgefs cluster. #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/edgefs/cluster/target/layout.go:
```diff
@@ -30,9 +30,9 @@ import (
+	// DefaultContainerMaxCapacity - max allowed container disks capacity, if exeeded then new new container will be added #nosec
 	DefaultContainerMaxCapacity = "132Ti"
-	S3PayloadSecretsPath        = "/opt/nedge/etc/secrets/"
+	S3PayloadSecretsPath        = "/opt/nedge/etc/secrets/" // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/edgefs/cluster/target/pod.go:
```diff
@@ -14,6 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// #nosec
```
`#nosec` on an entire package?!

BlaineEXE commented on pkg/operator/edgefs/cluster/utils.go:
```diff
@@ -281,7 +281,7 @@ func (c *cluster) resolveNode(nodeName string) *rookv1alpha2.Node {
 
 func (c *cluster) LabelTargetNode(nodeName string) {
-	c.AddLabelsToNode(nodeName, map[string]string{c.Namespace: "cluster"})
+	c.AddLabelsToNode(nodeName, map[string]string{c.Namespace: "cluster"}) // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/edgefs/cluster/utils.go:
```diff
@@ -307,7 +307,7 @@ func (c *cluster) AddLabelsToNode(nodeName string, labels map[string]string) err
 
 func (c *cluster) UnlabelTargetNode(nodeName string) {
-	c.RemoveLabelOffNode(nodeName, []string{c.Namespace})
+	c.RemoveLabelOffNode(nodeName, []string{c.Namespace}) // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/edgefs/s3x/s3x.go:
```diff
@@ -229,7 +229,7 @@ func (c *S3XController) makeDeployment(svcname, namespace, rookImage string, s3x
 		podSpec.Spec.DNSPolicy = v1.DNSClusterFirstWithHostNet
 	} else if c.NetworkSpec.IsMultus() {
-		k8sutil.ApplyMultus(c.NetworkSpec, &podSpec.ObjectMeta)
+		k8sutil.ApplyMultus(c.NetworkSpec, &podSpec.ObjectMeta) // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/k8sutil/k8sutil.go:
```diff
@@ -18,7 +18,7 @@ limitations under the License.
 
 import (
-	"crypto/md5"
+	"crypto/md5" // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/k8sutil/k8sutil.go:
```diff
@@ -75,7 +75,7 @@ func GetK8SVersion(clientset kubernetes.Interface) (*version.Version, error) {
 }
 
-// Hash MD5 hash a given string
+// Hash MD5 hash a given string #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/operator/nfs/server.go:
```diff
@@ -66,7 +66,7 @@ func Setup(ganeshaConfig string) error {
 	// Start ganesha.nfsd
 	logger.Infof("Running NFS server!")
-	cmd := exec.Command("ganesha.nfsd", "-F", "-L", ganeshaLog, "-f", ganeshaConfig, "-N", ganeshaOptions)
+	cmd := exec.Command("ganesha.nfsd", "-F", "-L", ganeshaLog, "-f", ganeshaConfig, "-N", ganeshaOptions) // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/util/exec/exec.go:
```diff
@@ -13,6 +13,8 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 limitations under the License.
 */
+
+// #nosec
```
`#nosec` on entire package!?

BlaineEXE commented on pkg/util/flags/flags.go:
```diff
@@ -13,6 +13,7 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
+// #nosec
```
`#nosec` on entire package!?

BlaineEXE commented on pkg/util/proc/procmanager.go:
```diff
@@ -123,7 +123,7 @@ func (p *ProcManager) Start(name, command, procSearchPattern string, policy Proc
 	p.RLock()
 	for _, proc := range p.procs {
-		proc.Stop(false)
+		proc.Stop(false) // #nosec
```
Ditto on document `#nosec`

BlaineEXE commented on pkg/util/sys/device.go:
```diff
@@ -313,7 +313,7 @@ func MountDeviceWithOptions(devicePath, mountPath, fstype, options string, execu
 	args = append(args, devicePath, mountPath)
 
-	os.MkdirAll(mountPath, 0755)
+	os.MkdirAll(mountPath, 0755) // #nosec
```
Ditto on document `#nosec`

nizamial09 commented:
> There are so many instances of `#nosec`, and I'm troubled. All uses should be documented with the error that exists and why it is okay to ignore the error in this case.
> 
> If you are unsure about what to do about a particular security error, please leave the error present, and do not merely ignore it with `#nosec`. It seems like many errors, especially errors on methods/functions or packages as a whole are ignored in these changes merely for convenience, which is not good practice. It turns a blind eye from real issues. Errors should be left in these cases so that the right people can fix them.
> 
> I am strongly against using `#nosec` on anything other than single lines in the code so that potential problems are not ignored in the future when developers add code to a function or package.

@BlaineEXE Okay. I guess you are right. I will add some comments on the comments. Most of the checks I used nosec is because of the variable name and some of them were like, the function expects for a certain directory permission and the directory permission used is not matched. For instance, some directory function expects for a value of 0750 as directory permission where the permission is set to 0755. So in that instance due to security issue this error shows. Changing the value to 0750 will invoke a different set of permissions, so that's not an option, I guess. That's why I put nosec in those instances.

nizamial09 commented:
@travisn @BlaineEXE @leseb 
Removed the usage of nosec from everywhere. 
For the directory permission security check, I've added a comment of the permissions. Please see to that if its okay. I've changed the permissions to the 'most narrow permission'. 

Unhandled errors are handled. Although I left out some warnings which I will explain below.

- **G204 (CWE-78): Subprocess launched with variable (Confidence: HIGH, Severity: MEDIUM)**
There are about 7 of these warnings. These are occurring when a CMD is called from the golang.
```It may not necessarily be an issue that introduces command injection but you may want to know which external executables are being called by an application when performing a security review.```

- **G501 (CWE-327): Blacklisted import crypto/md5: weak cryptographic primitive (Confidence: HIGH, Severity: MEDIUM)**
MD5 is cryptographically broken and should not be used for secure applications. Hence this warning. **I think this needs to be discussed more.**
Its used in this package: `pkg/operator/k8sutil/k8sutil.go`

- **G401 (CWE-326): Use of weak cryptographic primitive (Confidence: HIGH, Severity: MEDIUM)**
Same as above.

leseb commented on pkg/daemon/ceph/agent/agent.go:
```diff
@@ -113,7 +113,9 @@ func (a *Agent) Run() error {
 	stopChan := make(chan struct{})
-	clusterController.StartWatch(v1.NamespaceAll, stopChan)
+	if err := clusterController.StartWatch(v1.NamespaceAll, stopChan); err != nil {
+		logger.Errorf("Cluster Controller failed to start watch %+v", err)
```
```suggestion
		logger.Errorf("Cluster Controller failed to start watch. %v", err)
```

leseb replied:
This hasn't been addressed.

leseb commented on pkg/daemon/ceph/agent/flexvolume/server.go:
```diff
@@ -89,7 +90,9 @@ func (s *FlexvolumeServer) Start(driverVendor, driverName string) error {
 		logger.Info("Deleting unix domain socket file.")
-		os.Remove(unixSocketFile)
+		if err := os.Remove(unixSocketFile); err != nil {
+			logger.Errorf("failed to remove unix socket file- %+v", err)
```
```suggestion
			logger.Errorf("failed to remove unix socket file. %v", err)
```

leseb commented on pkg/daemon/ceph/agent/flexvolume/server.go:
```diff
@@ -98,7 +101,7 @@ func (s *FlexvolumeServer) Start(driverVendor, driverName string) error {
 	s.listeners[unixSocketFile] = listener
 
-	if err := os.Chmod(unixSocketFile, 0770); err != nil {
+	if err := os.Chmod(unixSocketFile, 0600); err != nil { // directory permission is 'drw-------'
```
I think you can remove the comment.

leseb commented on pkg/daemon/ceph/agent/flexvolume/server.go:
```diff
@@ -118,8 +121,10 @@ func (s *FlexvolumeServer) StopAll() {
 		if _, err := os.Stat(unixSocketFile); !os.IsNotExist(err) {
-			logger.Infof("deleting unix domain socket file %q.", unixSocketFile)
-			os.Remove(unixSocketFile)
+			logger.Infof("Deleting unix domain socket file %s.", unixSocketFile)
```
```suggestion
			logger.Infof("Deleting unix domain socket file %q.", unixSocketFile)
```

leseb commented on pkg/daemon/ceph/agent/flexvolume/server.go:
```diff
@@ -118,8 +121,10 @@ func (s *FlexvolumeServer) StopAll() {
-			os.Remove(unixSocketFile)
+			logger.Infof("Deleting unix domain socket file %s.", unixSocketFile)
+			if err := os.Remove(unixSocketFile); err != nil {
+				logger.Errorf("failed to delete unix domain socker file- %+v", err)
```
```suggestion
				logger.Errorf("failed to delete unix domain socket file. %v", err)
```

leseb commented on pkg/daemon/ceph/agent/flexvolume/server.go:
```diff
@@ -193,7 +198,7 @@ func LoadFlexSettings(directory string) []byte {
 	// copying flex volume
 	if _, err := os.Stat(driverDir); os.IsNotExist(err) {
-		err := os.Mkdir(driverDir, 0755)
+		err := os.Mkdir(driverDir, 0750) // directory permission set to 750 which restrict the world from 'rwx' and group from 'w'
```
I think you can remove the comment.

leseb commented on pkg/daemon/ceph/agent/flexvolume/server.go:
```diff
@@ -246,13 +251,13 @@ func configureFlexVolume(driverFile, driverDir, driverName string) error {
 	defer srcFile.Close()
 
-	destFile, err := os.OpenFile(dest, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0755) // creates if file doesn't exist
+	destFile, err := os.OpenFile(dest, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0600) // creates if file doesn't exist, permission is 'drw-------'
```
Unnecessary additional comment.

leseb commented on pkg/daemon/ceph/osd/agent.go:
```diff
@@ -461,7 +461,9 @@ func (a *OsdAgent) prepareOSD(context *clusterd.Context, cfg *osdConfig) (*oposd
 	if devPartInfo != nil {
-		sys.UnmountDevice(devPartInfo.pathToUnmount, context.Executor)
+		if err := sys.UnmountDevice(devPartInfo.pathToUnmount, context.Executor); err != nil {
+			logger.Errorf("Failed to Unmount Device - %+v", err)
```
```suggestion
			logger.Errorf("Failed to unmount device. %v", err)
```

leseb commented on pkg/daemon/ceph/osd/daemon.go:
```diff
@@ -46,7 +46,7 @@ func StartOSD(context *clusterd.Context, osdType, osdID, osdUUID, lvPath string,
 	// ensure the config mount point exists
 	configDir := fmt.Sprintf("/var/lib/ceph/osd/ceph-%s", osdID)
-	err := os.Mkdir(configDir, 0755)
+	err := os.Mkdir(configDir, 0750) // directory permission is drwxr-x---
```
ditto

leseb commented on pkg/daemon/ceph/test/info.go:
```diff
@@ -1,13 +1,24 @@
 )
 
+var (
+	logger = capnslog.NewPackageLogger("github.com/rook/rook", "cephosd")
```
cephosd? The name should be different.

nizamial09 replied:
Oops. My bad.

leseb commented on pkg/daemon/util/copybins.go:
```diff
@@ -67,5 +68,5 @@ func copyBinary(sourceDir, targetDir, filename string) error {
 	}
 
-	return os.Chmod(targetPath, 0755)
+	return os.Chmod(targetPath, 0600) // directory permission is 'drw-------'
```
ditto

leseb commented on pkg/operator/ceph/cluster/crash/crash.go:
```diff
@@ -46,7 +46,7 @@ import (
 const (
 	crashCollectorKeyringUsername = "client.crash"
-	crashCollectorSecretName      = "rook-ceph-crash-collector-keyring"
+	crashCollectorScrtName        = "rook-ceph-crash-collector-keyring"
```
why?

nizamial09 replied:
Apparently Secret is a keyword that gets triggered by gosec. 

Variables are considered to look like a password if they have match any one of:

```- “password”
- “pass”
- “passwd”
- “pwd”
- “secret”
- “token”```

Note: this can be noisy and may generate false positives.

nizamial09 replied:
@leseb

leseb commented on pkg/operator/ceph/cluster/mgr/dashboard.go:
```diff
@@ -39,7 +39,7 @@ const (
 	dashboardPortHTTP              = 7000
 	dashboardUsername              = "admin"
-	dashboardPasswordName          = "rook-ceph-dashboard-password"
+	dashboardPdName                = "rook-ceph-dashboard-password"
```
Why?

nizamial09 replied:
The word Password is trigerring the gosec. I changed the variable name to `dashboardPwdName` and yet it trigerred the gosec as a security issue. A better variable name would be what? @leseb

BlaineEXE replied:
I think it's okay to use `#nosec for this specific line and comment that gosec flags the variable as a potential problem because it has the word `password` in it. We aren't actually using the password, and we know it, so this is safe to ignore. Otherwise devs who come later won't know what `Pd` means.

leseb commented on pkg/operator/ceph/csi/secrets.go:
```diff
@@ -28,15 +28,15 @@ import (
 	csiKeyringRBDNodeUsername        = "client.csi-rbd-node"
-	csiRBDNodeSecret                 = "rook-csi-rbd-node"
-	csiRBDProvisionerSecret          = "rook-csi-rbd-provisioner"
+	csiRBDNodeScrt                   = "rook-csi-rbd-node"
```
ditto

leseb commented on pkg/util/proc/procmanager.go:
```diff
@@ -120,13 +120,16 @@ func (p *ProcManager) Start(name, command, procSearchPattern string, policy Proc
 	for _, proc := range p.procs {
-		proc.Stop(false)
+		if err := proc.Stop(false); err != nil {
+			return fmt.Errorf("error while shutting down - %+v", err)
```
```suggestion
			return fmt.Errorf("error while shutting down. %v", err)
```

leseb commented on pkg/util/sys/device.go:
```diff
@@ -313,7 +313,9 @@ func MountDeviceWithOptions(devicePath, mountPath, fstype, options string, execu
 
-	os.MkdirAll(mountPath, 0755)
+	if err := os.MkdirAll(mountPath, 0750); err != nil { //directory permission is 'drwxr-x---'
+		return fmt.Errorf("make directory failed - %+v", err)
```
ditto

leseb commented on pkg/util/flags/flags.go:
```diff
@@ -82,24 +82,35 @@ func createRequiredFlagError(name string, flags []string) error {
 
 func SetFlagsFromEnv(flags *pflag.FlagSet, prefix string) error {
+	errorFlag := false
+	var err error
```
You don't need to declare it, simply use `:=` below.

nizamial09 replied:
flags.VisitAll(.....) is a function and inside that function this `errorFlag` is being set to true. This function doesn't support a return call. That's why I initialized it at the beginning. The `errorFlag != false` is being checkout outside of the `VisitAll()` function. @leseb

nizamial09 replied:
I can make it 
```var (
       errorFlag bool
       err error
  )
```

BlaineEXE replied:
I think Seb is referring to `var err error`

nizamial09 replied:
Ah yes. That makes sense.Sure. I'll do the changes.

nizamial09 replied:
Oh. I cannot make that change also. Because the error is also displayed outside this `VisitAll()` function.

nizamial09 commented:
@leseb Ammended the changes. Updated the commit description. Replied to your comments. Please check it out.

nizamial09 commented:
> Put the whole description of the PR in your commit and try to make it aesthetic... Read https://www.conventionalcommits.org/en/v1.0.0-beta.2/ if necessary.
> Also include the issues you are fixing in your commit message.
> 
> I'm updating the commit guideline here in the meantime... #4605

Thanks. I've edited the commit message. I hope it explains most about the PR. @leseb

leseb commented:
> > Put the whole description of the PR in your commit and try to make it aesthetic... Read https://www.conventionalcommits.org/en/v1.0.0-beta.2/ if necessary.
> > Also include the issues you are fixing in your commit message.
> > I'm updating the commit guideline here in the meantime... #4605
> 
> Thanks. I've edited the commit message. I hope it explains most about the PR. @leseb

The commit looks much better thanks, just add an empty line after the commit title :).

nizamial09 commented:
> > > Put the whole description of the PR in your commit and try to make it aesthetic... Read https://www.conventionalcommits.org/en/v1.0.0-beta.2/ if necessary.
> > > Also include the issues you are fixing in your commit message.
> > > I'm updating the commit guideline here in the meantime... #4605
> > 
> > 
> > Thanks. I've edited the commit message. I hope it explains most about the PR. @leseb
> 
> The commit looks much better thanks, just add an empty line after the commit title :).

Done. ;) Thanks :)

leseb commented:
> > > > Put the whole description of the PR in your commit and try to make it aesthetic... Read https://www.conventionalcommits.org/en/v1.0.0-beta.2/ if necessary.
> > > > Also include the issues you are fixing in your commit message.
> > > > I'm updating the commit guideline here in the meantime... #4605
> > > 
> > > 
> > > Thanks. I've edited the commit message. I hope it explains most about the PR. @leseb
> > 
> > 
> > The commit looks much better thanks, just add an empty line after the commit title :).
> 
> Done. ;) Thanks :)

I don't see the empty line after the commit title.

nizamial09 commented:
> > > > > Put the whole description of the PR in your commit and try to make it aesthetic... Read https://www.conventionalcommits.org/en/v1.0.0-beta.2/ if necessary.
> > > > > Also include the issues you are fixing in your commit message.
> > > > > I'm updating the commit guideline here in the meantime... #4605
> > > > 
> > > > 
> > > > Thanks. I've edited the commit message. I hope it explains most about the PR. @leseb
> > > 
> > > 
> > > The commit looks much better thanks, just add an empty line after the commit title :).
> > 
> > 
> > Done. ;) Thanks :)
> 
> I don't see the empty line after the commit title.
See this image. This is the one right? I put an empty space while amending it.

![commit](https://user-images.githubusercontent.com/53651462/71736038-14e82c00-2e76-11ea-8756-43ccbc365c0a.png)

leseb commented:
> > > > > > Put the whole description of the PR in your commit and try to make it aesthetic... Read https://www.conventionalcommits.org/en/v1.0.0-beta.2/ if necessary.
> > > > > > Also include the issues you are fixing in your commit message.
> > > > > > I'm updating the commit guideline here in the meantime... #4605
> > > > > 
> > > > > 
> > > > > Thanks. I've edited the commit message. I hope it explains most about the PR. @leseb
> > > > 
> > > > 
> > > > The commit looks much better thanks, just add an empty line after the commit title :).
> > > 
> > > 
> > > Done. ;) Thanks :)
> > 
> > 
> > I don't see the empty line after the commit title.
> > See this image. This is the one right? I put an empty space while amending it.
> 
> ![commit](https://user-images.githubusercontent.com/53651462/71736038-14e82c00-2e76-11ea-8756-43ccbc365c0a.png)

Oh ok, it looks like a github issue. Looks good.

leseb commented on pkg/operator/edgefs/cluster/mgr/mgr.go:
```diff
@@ -344,7 +344,9 @@ func (c *Cluster) makeDeployment(name, clusterName, rookImage string, replicas i
 	} else if c.NetworkSpec.IsMultus() {
-		k8sutil.ApplyMultus(c.NetworkSpec, &podSpec.ObjectMeta)
+		if err := k8sutil.ApplyMultus(c.NetworkSpec, &podSpec.ObjectMeta); err != nil {
+			logger.Errorf("apply multus failed- %+v", err)
```
ditto

nizamial09 replied:
Oops. Missed it. I'll clear it up.

BlaineEXE commented:
I really think with this many disparate changes, there should be different commits. At least a separate commit for each of the different `G` types mentioned in the description.

```
Fixed G101: Potential hardcoded credential variable by renaming certain variables
Fixed G104: Unhandled Errors are handled now
Fixed G303: Poor file permissions used for directory by using the most narrow permission
Fixed G304: Potential File inclusion by cleaning up the file path
```

nizamial09 commented:
> I really think with this many disparate changes, there should be different commits. At least a separate commit for each of the different `G` types mentioned in the description.
> 
> ```
> Fixed G101: Potential hardcoded credential variable by renaming certain variables
> Fixed G104: Unhandled Errors are handled now
> Fixed G303: Poor file permissions used for directory by using the most narrow permission
> Fixed G304: Potential File inclusion by cleaning up the file path
> ```

Oh. Is it possible to do now, since I made all of these changes into a single commit?

nizamial09 commented:
@leseb @BlaineEXE @travisn Fixed all the errors. Split it into multiple commits.

nizamial09 commented:
Updated the description with latest GOSEC summary report.

BlaineEXE commented on pkg/daemon/ceph/test/info.go:
```diff
@@ -1,13 +1,24 @@
+	logger = capnslog.NewPackageLogger("github.com/rook/rook", "cephtest")
+)
+
 func CreateConfigDir(configDir string) {
```
These methods are for testing right? I wonder if it is best to panic if we find an error here rather than just printing an error. Printing the error isn't likely to get the user's attention, and it won't change the results of the test. @travisn @galexrt ?

nizamial09 replied:
Yes. That is true. I'll return this error and give appropriate checks from the _test.go files.

BlaineEXE commented on pkg/operator/edgefs/cluster/target/layout.go:
```diff
@@ -32,7 +32,7 @@ import (
 	// DefaultContainerMaxCapacity - max allowed container disks capacity, if exeeded then new new container will be added
 	DefaultContainerMaxCapacity = "132Ti"
-	S3PayloadSecretsPath        = "/opt/nedge/etc/secrets/"
+	S3PayloadSecretsPath        = "/opt/nedge/etc/secrets/" // #nosec since gosec flags the variable as a potential problem because of the word `Secret`
```
Move to above the code line?

BlaineEXE commented on pkg/operator/edgefs/cluster/mgr/mgr.go:
```diff
@@ -344,7 +344,9 @@ func (c *Cluster) makeDeployment(name, clusterName, rookImage string, replicas i
 	} else if c.NetworkSpec.IsMultus() {
-		k8sutil.ApplyMultus(c.NetworkSpec, &podSpec.ObjectMeta)
+		if err := k8sutil.ApplyMultus(c.NetworkSpec, &podSpec.ObjectMeta); err != nil {
+			logger.Errorf("apply multus failed. %v", err)
```
More context.

```suggestion
			logger.Errorf("failed to apply multus spec to pod template metadata. %v", err)
```

BlaineEXE commented on pkg/operator/edgefs/cluster/controller.go:
```diff
@@ -225,7 +233,9 @@ func (c *ClusterController) onAdd(obj interface{}) {
 		cluster.Spec.UseHostLocalTime)
-	ISCSIController.StartWatch(cluster.stopCh)
+	if err := ISCSIController.StartWatch(cluster.stopCh); err != nil {
+		logger.Errorf("error while watching ISCSI service CRD. %v", err)
```
```suggestion
		logger.Errorf("failed to start watch for edgefs watching ISCSI service. %v", err)
```

BlaineEXE commented on pkg/operator/edgefs/cluster/controller.go:
```diff
@@ -238,7 +248,9 @@ func (c *ClusterController) onAdd(obj interface{}) {
 		cluster.Spec.UseHostLocalTime)
-	ISGWController.StartWatch(cluster.stopCh)
+	if err := ISGWController.StartWatch(cluster.stopCh); err != nil {
+		logger.Errorf("error while watching ISGW service CRD. %v", err)
```
```suggestion
		logger.Errorf("failed to start watch for ISGW service. %v", err)
```

BlaineEXE commented on pkg/operator/edgefs/cluster/controller.go:
```diff
@@ -212,7 +218,9 @@ func (c *ClusterController) onAdd(obj interface{}) {
 		cluster.Spec.UseHostLocalTime)
-	S3XController.StartWatch(cluster.stopCh)
+	if err := S3XController.StartWatch(cluster.stopCh); err != nil {
+		logger.Errorf("error while watching s3xcontroller service CRD. %v", err)
```
```suggestion
		logger.Errorf("failed to start watch for s3x controller service. %v", err)
```

BlaineEXE commented on pkg/operator/edgefs/cluster/controller.go:
```diff
@@ -199,7 +203,9 @@ func (c *ClusterController) onAdd(obj interface{}) {
 		cluster.Spec.UseHostLocalTime)
-	SWIFTController.StartWatch(cluster.stopCh)
+	if err := SWIFTController.StartWatch(cluster.stopCh); err != nil {
+		logger.Errorf("error while watching SWIFT service CRD. %v", err)
```
```suggestion
		logger.Errorf("failed to start watch for SWIFT service. %v", err)
```

BlaineEXE commented on pkg/operator/edgefs/cluster/controller.go:
```diff
@@ -186,7 +188,9 @@ func (c *ClusterController) onAdd(obj interface{}) {
 		cluster.Spec.UseHostLocalTime)
-	S3Controller.StartWatch(cluster.stopCh)
+	if err := S3Controller.StartWatch(cluster.stopCh); err != nil {
+		logger.Errorf("error while watching s3 service CRD. %v", err)
```
```suggestion
		logger.Errorf("failed to start watch for s3 service. %v", err)
```

BlaineEXE commented on pkg/operator/edgefs/cluster/controller.go:
```diff
@@ -173,7 +173,9 @@ func (c *ClusterController) onAdd(obj interface{}) {
 		cluster.Spec.UseHostLocalTime)
-	NFSController.StartWatch(cluster.stopCh)
+	if err := NFSController.StartWatch(cluster.stopCh); err != nil {
+		logger.Errorf("error while watching nfs service CRD. %v", err)
```
```suggestion
		logger.Errorf("failed to start watch for nfs service. %v", err)
```

BlaineEXE commented on pkg/operator/cockroachdb/operator.go:
```diff
@@ -58,7 +58,9 @@ func (o *Operator) Run() error {
 	// watch for changes to the cockroachdb clusters
-	o.clusterController.StartWatch(v1.NamespaceAll, stopChan)
+	if err := o.clusterController.StartWatch(v1.NamespaceAll, stopChan); err != nil {
+		logger.Errorf("start watch failed for cockroachdb clusters. %v", err)
```
```suggestion
		logger.Errorf("failed to start watch for cockroachdb clusters. %v", err)
```

BlaineEXE commented on pkg/operator/ceph/test/podspec.go:
```diff
@@ -67,9 +67,9 @@ func NewPodSpecTester(t *testing.T, spec *v1.PodSpec) *PodSpecTester {
+	keyringSecretName := fmt.Sprintf("rook-ceph-%s-%s-keyring", daemonType, daemonID) // #nosec since gosec flags the variable as a potential problem because of the word `Secret`
 	if daemonType == config.MonType {
-		keyringSecretName = "rook-ceph-mons-keyring" // mons share a keyring
+		keyringSecretName = "rook-ceph-mons-keyring" // mons share a keyring #nosec since gosec flags the variable as a potential problem because of the word `Secret`
```
`nosec` line above?

BlaineEXE commented on pkg/operator/ceph/test/podspec.go:
```diff
@@ -67,9 +67,9 @@ func NewPodSpecTester(t *testing.T, spec *v1.PodSpec) *PodSpecTester {
 	daemonType config.DaemonType, daemonID string,
 ) {
-	keyringSecretName := fmt.Sprintf("rook-ceph-%s-%s-keyring", daemonType, daemonID)
+	keyringSecretName := fmt.Sprintf("rook-ceph-%s-%s-keyring", daemonType, daemonID) // #nosec since gosec flags the variable as a potential problem because of the word `Secret`
```
`nosec` line above?

BlaineEXE commented on pkg/operator/ceph/operator.go:
```diff
@@ -149,7 +149,9 @@ func (o *Operator) Run() error {
 	// watch for changes to the rook clusters
-	o.clusterController.StartWatch(namespaceToWatch, stopChan)
+	if err := o.clusterController.StartWatch(namespaceToWatch, stopChan); err != nil {
+		logger.Errorf("start watch failed for rook clusters. %v", err)
```
```suggestion
		logger.Errorf("failed to start watch for Ceph clusters. %v", err)
```

nizamial09 commented on pkg/daemon/discover/discover.go:
```diff
@@ -388,7 +391,7 @@ func updateDeviceCM(context *clusterd.Context) error {
 	devices := make([]sys.LocalDisk, 0)
 	localDevices, err := clusterd.DiscoverDevices(context.Executor)
-	logger.Infof("localdevices - %+v", localDevices)
+	logger.Infof("localdevices. %v", localDevices)
```
Is there any place I could go and refer for these standard syntaxes? @BlaineEXE @leseb @travisn

leseb commented on pkg/operator/cassandra/sidecar/config.go:
```diff
@@ -417,8 +417,12 @@ func getJolokiaConfig() string {
-	yaml.Unmarshal(initialYAML, &initial)
-	yaml.Unmarshal(overrideYAML, &override)
+	if err := yaml.Unmarshal(initialYAML, &initial); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal initial yaml. %v", err)
```
`Wrapf` is not implemented in cassandra.

nizamial09 commented:
> @nizamial09 Could you resolve the conflicts and review the next round of feedback? It would be great to get this merged soon, thanks!

Sorry for the delay. Will push in the by today.

nizamial09 commented on pkg/util/flags/flags.go:
```diff
@@ -82,24 +82,37 @@ func createRequiredFlagError(name string, flags []string) error {
 			// Set the environment variable. Will override default values, but be overridden by command line parameters.
-			flags.Set(f.Name, value)
+			if err = flags.Set(f.Name, value); err != nil {
+				errorFlag = true
```
Thank you

nizamial09 commented:
> What's the status on this PR? Thanks
@leseb 
The integration tests are mostly failing because of the permission settings as @travisn said. I tried the method mentioned by @travisn and was able to create and acces the mysql pod. I pinged travisn related to this and meanwhile I was doing some tests on my own.

leseb commented:
> > What's the status on this PR? Thanks
> 
> @leseb
> The integration tests are mostly failing because of the permission settings as @travisn said. I tried the method mentioned by @travisn and was able to create and acces the mysql pod. I pinged travisn related to this and meanwhile I was doing some tests on my own.

If it's working please rebase and see if it passes CI.

nizamial09 commented:
> > > What's the status on this PR? Thanks
> > 
> > 
> > @leseb
> > The integration tests are mostly failing because of the permission settings as @travisn said. I tried the method mentioned by @travisn and was able to create and acces the mysql pod. I pinged travisn related to this and meanwhile I was doing some tests on my own.
> 
> If it's working please rebase and see if it passes CI.

nizamial09 commented:
> > > > What's the status on this PR? Thanks
> > > 
> > > 
> > > @leseb
> > > The integration tests are mostly failing because of the permission settings as @travisn said. I tried the method mentioned by @travisn and was able to create and acces the mysql pod. I pinged travisn related to this and meanwhile I was doing some tests on my own.
> > 
> > 
> > If it's working please rebase and see if it passes CI.

I'll do that. But I am getting some error while I was testing this myself. So trying to figure it out @leseb

nizamial09 commented:
> @nizamial09 How about if you split this PR into four separate PRs, one for each of your commits? Then it will be easier to narrow down the permissions issues, and the other commits such as error checking can go ahead and get merged.

That would be much better. Then I'll split into multiple PR and will push it in by tomorrow and the next day, is that okay? @travisn

nizamial09 commented:
Splitted this into 4 PR's
https://github.com/rook/rook/pull/4865
https://github.com/rook/rook/pull/4866
https://github.com/rook/rook/pull/4867
https://github.com/rook/rook/pull/4868
